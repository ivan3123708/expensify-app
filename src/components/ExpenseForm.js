import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment().format('DD MMM YYYY');
console.log(now);

export default class ExpenseForm extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      note: props.expense ? props.expense.note : '',
      focused: false,
      error: ''
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description: description}))
  }

  onAmountChange = (e) => {
    const amount = e.target.value;

  if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({amount: amount}))
    }
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({note: note}))
  }

  onDateChange = (createdAt) => {
    if(createdAt) {
      this.setState(() => ({createdAt: createdAt}));
    }
  }

  onFocusChange = ({focused}) => {
    this.setState(() => ({focused: focused}));
  }

  onSubmit = (e) => {
    e.preventDefault();

    if(!this.state.description || !this.state.amount) {
      this.setState(() => ({error: 'Please enter description and amount.'}));
    } else {
      this.setState(() => ({error: ''}));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea 
            placeholder="Add note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
}