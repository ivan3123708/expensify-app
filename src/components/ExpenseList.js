import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getFilteredExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    <h3>Expense List</h3>
    {props.expenses.length > 0 ? (
      props.expenses.map((expense) => {
        return <ExpenseListItem key={expense.id} {...expense}/>
      })) : (
        <p>No expenses</p>
      )
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: getFilteredExpenses(state.expenses, state.filters)
  }
};

export default connect(mapStateToProps)(ExpenseList);