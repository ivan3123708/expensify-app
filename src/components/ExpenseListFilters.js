import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {

  state = {
    focusedInput: null
  }

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }

  onSortChange = (e) => {
    e.target.value === "date" ? this.props.sortByDate() : this.props.sortByAmount();
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  onFocusChange = (focusedInput) => {
    this.setState(() => ({ focusedInput: focusedInput }));
  }

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focusedInput}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    )
  }
};

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);