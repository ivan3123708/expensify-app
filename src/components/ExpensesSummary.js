import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import { getFilteredExpenses, getSummedExpenses } from '../selectors/expenses';

export const ExpensesSummary = (props) => {
  const singular_plural = props.count === 1 ? 'expense' : 'expenses';
  return (
    <div>
      <p>Viewing {props.count} {singular_plural} totalling {numeral(props.sum / 100).format('$0,0.00')}</p>
    </div>
  )
};

const mapStateToProps = (state) => ({
  count: getFilteredExpenses(state.expenses, state.filters).length,
  sum: getSummedExpenses(getFilteredExpenses(state.expenses, state.filters))
});

export default connect(mapStateToProps)(ExpensesSummary);