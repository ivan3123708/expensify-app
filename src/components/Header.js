import React from 'react';
import {NavLink} from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" activeClassName="selected" exact>Dashboard</NavLink>
        <NavLink to="/create" activeClassName="selected">Create Expense</NavLink>
    </header>
);

export default Header;