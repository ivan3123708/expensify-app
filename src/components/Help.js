import React from 'react';

const Help = () => (
    <div>
        This is help page
    </div>
);

export default Help;