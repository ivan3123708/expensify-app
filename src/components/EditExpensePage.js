import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class Edit extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }

  onClick = (expense) => {
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}/>
        <button onClick={this.onClick}>Remove</button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id;
    })
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense(id))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Edit);