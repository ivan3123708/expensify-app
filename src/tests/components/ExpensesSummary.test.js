import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('Render ExpensesSummary with single expense', () => {
  const result = shallow(<ExpensesSummary count={1} sum={150} />);
  expect(result).toMatchSnapshot();
});

test('Render ExpensesSummary with multiple expenses', () => {
  const result = shallow(<ExpensesSummary count={4} sum={1800000} />);
  expect(result).toMatchSnapshot();
});