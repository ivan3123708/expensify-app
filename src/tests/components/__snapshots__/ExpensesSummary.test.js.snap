// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ExpensesSummary with multiple expenses 1`] = `
<div>
  <p>
    Viewing 
    4
     
    expenses
     totalling 
    $18,000.00
  </p>
</div>
`;

exports[`Render ExpensesSummary with single expense 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    $1.50
  </p>
</div>
`;
